File: code/ch18/18.6.2.java
public int countPrimes(int n) {
    if (n <= 1) {
        return 0;
    }
    boolean[] tmp = new boolean[n];
    Arrays.fill(tmp, true);
    int ans = 0;
    tmp[0] = false;
    tmp[1] = false;
    for (int i = 2; i < n; i++) {
        if (!tmp[i]) {
            continue;
        }
        // 查找 i 倍数的优化细节：i * i开始
        for (int j = i * i; j < n; j += i) {
            tmp[j] = false;
        }
    }
    for (int i = 0; i < n; i++) {
        if (tmp[i]) {
            ans += 1;
        }
    }
    return ans;
}

File: code/ch18/18.6.2.py
class Solution:
    def countPrimes(self, n: int) -> int:
        if n <= 1:
            return 0
        tmp = [True for i in range(n)]
        ans = 0
        tmp[0] = False
        tmp[1] = False
        for i in range(2, n):
            if not tmp[i]:
                continue
            # 查找 i 倍数的优化细节：i * i开始
            for j in range(i * i, n, i):
                tmp[j] = False
        for i in range(n):
            if tmp[i]:
                ans += 1
        return ans
